### PR TITLE
kraken: Remove a few use-after-free error

### PR DESCRIPTION
**Pull request by kinnou02**
@nonifier suggested doing some check with https://github.com/google/sanitizers/wiki/AddressSanitizer
Not it's best idea of the year...

Kraken had one in `apply_disruption()` (I'am sure everyone is shocked), I have put a band aid on it...

The second commit fix a few things in the tests.

I have run a kraken with realtime feeding, there was no error nor leak detected.
A lot of tests leak but it isn't a surprise as we allocate a lot of fixtures on the heap without even trying to free it properly. I will try to clean this shit to allow us to run test with the sanitizers on.

### CODE REVIEW

File: source/kraken/apply_disruption.cpp [hunk 1]
```diff
@@ -54,6 +54,47 @@ namespace bg = boost::gregorian;
 
 namespace {
 
+static nt::VehicleJourney* create_vj_from_old_vj(nt::MetaVehicleJourney* mvj,
+                              const nt::VehicleJourney* vj,
+                              const std::string& new_vj_uri,
+                              nt::RTLevel rt_level,
+                              nt::ValidityPattern new_vp,
+                              const std::vector<nt::StopTime>& new_stop_times,
```
Comment (TeXitoi) [inline source/kraken/apply_disruption.cpp]: you want to take that by value to have a useful `stdd::move` usage.

File: source/kraken/apply_disruption.cpp [hunk 2]
```diff
@@ -54,6 +54,47 @@ namespace bg = boost::gregorian;
...
+    auto odt_message = vj->odt_message;
+    auto vehicle_properties = vj->_vehicle_properties;
+
+    auto* new_vj = mvj->create_discrete_vj(new_vj_uri,
+            rt_level,
+            new_vp,
+            vj->route,
+            std::move(new_stop_times),
+            pt_data);
+    vj = nullptr;
```
Comment (TeXitoi) [inline source/kraken/apply_disruption.cpp]: why?

**Review by TeXitoi — CHANGES_REQUESTED** [review]

File: source/kraken/apply_disruption.cpp (hunk 2 above)
Comment (kinnou02) [reply source/kraken/apply_disruption.cpp]: `create_discrete_vj()` also triger the deletion of useless vjs, so when a disruption is applied on an already adapte/realtime vj, the *base* vj is deleted when we create the new one.
To be safe I set `vj = nullptr` to explicitly indicate that we can't use it anymore.

Comment (TeXitoi) [reply source/kraken/apply_disruption.cpp]: A comment explaining why it is set to null would be great.

File: source/kraken/apply_disruption.cpp [hunk 3]
```diff
@@ -54,6 +54,47 @@ namespace bg = boost::gregorian;
 
 namespace {
 
+static nt::VehicleJourney* create_vj_from_old_vj(nt::MetaVehicleJourney* mvj,
+                              const nt::VehicleJourney* vj,
+                              const std::string& new_vj_uri,
+                              nt::RTLevel rt_level,
+                              nt::ValidityPattern new_vp,
+                              const std::vector<nt::StopTime> new_stop_times,
```
Comment (TeXitoi) [inline source/kraken/apply_disruption.cpp]: It will copy... You have to remove const.

**Review by TeXitoi — CHANGES_REQUESTED** [review]

File: source/kraken/apply_disruption.cpp [hunk 4]
```diff
@@ -442,32 +470,18 @@ struct add_impacts_visitor : public apply_impacts_visitor {
...
+            auto* new_vj = create_vj_from_old_vj(
+                    mvj,
+                    vj,
+                    new_vj_uri,
                     rt_level,
                     new_vp,
-                    vj->route,
                     std::move(new_stop_times),
                     pt_data);
+            vj = nullptr; //after the call to create_vj, vj may have been deleted :(
```
Comment (benoit-bst) [inline source/kraken/apply_disruption.cpp:482]: But, it is already done here ?
https://github.com/CanalTP/navitia/pull/2613/commits/312b1d6da9bba5f6e197247022ff8fcee326fa73#diff-c234f9d9583465696c5f3311313759d0R75

**Review by TeXitoi — APPROVED** [review]

File: source/kraken/apply_disruption.cpp (hunk 4 above)
Comment (TeXitoi) [reply source/kraken/apply_disruption.cpp:482]: That's local variables

**Review by benoit-bst — APPROVED** [review]

**Pull request merged**